service/storage_service.cc

```diff
@@ -354,6 +354,16 @@ future<> storage_service::topology_state_load(cdc::generation_service& cdc_gen_s
             // Save tokens, not needed for raft topology management, but needed by legacy
             // Also ip -> id mapping is needed for address map recreation on reboot
             if (!utils::fb_utilities::is_me(ip)) {
+                // We don't load gossiper endpoint states in storage_service::join_cluster
+                // if _raft_topology_change_enabled. On the other hand gossiper is still needed
+                // even in case of _raft_topology_change_enabled mode, since it still contains part
+                // of the cluster state. To work correctly, the gossiper needs to know the current
+                // endpoints. We cannot rely on seeds alone, since it is not guaranteed that seeds
+                // will be up to date and reachable at the time of restart.
+                if (!_gossiper.get_endpoint_state_for_endpoint_ptr(ip)) {
+                    co_await _gossiper.add_saved_endpoint(ip);
```

Comment:
We are in the middle of creation of the token metadata object here and `add_saved_endpoint` access the token metadata. Shouldn't this be done after the object is fully created otherwise it looks like we access the previous version of the object here.

Reply:
fixed